Refactor add tests

JMurphyWeb opened the pull request:
@mk4111 I don't know how to add comments to a branch so opened a PR.

- Small refactor handlers
- Add tests to the endpoints we worked on on friday 

~~(not passing when all run together but fine when run in isolation. Couldn't see why?)~~

# READY FOR REVIEW - PLEASE MERGE IN AFTER PR #82

mk4111 commented:
@JMurphyWeb, awesome!
Just had a look at it now, and have made some changes (hope you don't mind!) and fixed the issue.
Summary of changes made :
- changed the file names and some variable names so that they were a little more clear - I hope you don't mind...
- moved some of the tests into folders - in retrospect, this probably wasn't necessary, but as the number of tests increase I thought it might be nice to know where to go to find the relevant ones.
- switched the t.equal('check password matches') fn with t.ok(bcrypt.compareSync)
- changed/added to some of your comments.
- in handlers.js, changed your passing of the arguments as two seperate arguments. @JMurphyWeb was there a reason why you chose to do it your way? Am happy to revert back to your method but am curious...

Below are the changes made to fix the error :

```
{ AbortError: SOME_REDIS_FUNCTION can't be processed. The connection is already closed... }
```
where SOME_REDIS_FUNCTION was LINDEX, LLEN, etc.

(1) server.stop() was swapped out for process.exit(0). The issue faced seemed to be linked to server.stop() as this order acts to [reset the server state and refuse/close any new/existing connections](http://hapijs.com/api#serverstopoptions-callback). process.exit(0) on the other hand, when explicitly stated as we have done in our code, [terminates node processes](https://nodejs.org/api/process.html#process_process_exit_code). I'm not sure whether this is bad practise, but there was a [relevant issue](https://github.com/substack/tape/issues/216#issuecomment-224728386) which suggested this as a solution. In general though, @nelsonic, is this bad practise? What other solutions to this issue would you recommend?

(2) changes tape('teardown') fn to tape.onFinish --> on [tape docs](https://github.com/substack/tape/blob/e532790dd690ac8936d642b1392b14cba91e2569/readme.markdown#testonfinishfn) it mentions that this function only runs when _all_ the tests have finished, but from our use here, I think it runs when all the tests _in a file_ finish, as some tests clash if the database is not flushed in between them, and this has not happened.

Given that this pull-request is dependent on the user-creation-form being merged in (because you've made the PR to master), I'm going to assign this to @nelsonic - hope that's ok!

mk4111 commented:
@nelsonic, I'm not sure if I should just open up a new issue for this, because I'm still a little confused about why similar test setups might be working for some projects (e.g. ageuk's companion app), and why it's failing on others. I came across [this issue](https://github.com/hapijs/hapi/issues/2723) and although
I'm not sure how relevant it is, because they have specified some caching setup with Hapi, is it possible that this is issue is coming from server-side caching using catbox? Do we need to override this default with catbox-redis? A little confused, but I thought it was worth asking.

Happy to move this conversation into an issue, but not sure if making an issue on this repo is the most effective way of figuring out the cause of the problem...

JMurphyWeb commented:
Good stuff @mk4111 i will reply to your message tomorrow. Maybe we should close other PR and concentrate on getting this merged?

JMurphyWeb commented:
> - changed the file names and some variable names so that they were a little more clear - I hope you don't mind...
- moved some of the tests into folders - in retrospect, this probably wasn't necessary, but as the number of tests increase I thought it might be nice to know where to go to find the relevant ones.
- switched the t.equal('check password matches') fn with t.ok(bcrypt.compareSync)
- changed/added to some of your comments.

All sounds & looks good 👍 .

> - in handlers.js, changed your currying function. @JMurphyWeb was there a reason why you chose to do it your way? Am happy to revert back to your method but am curious...

@mk4111 Sorry that was hangover from previous project, I should have taken it out.  Our Redis functions are going to have a fair bit of repetition though, so maybe we can come back to this when we get round to refactoring?

JMurphyWeb commented:
@mk4111 one question about the new route names. I think you changed /users/ to  /people/ and i can see why, but I wonder if we should make an issue to check with Alex? 

I only say that because I think /people/ sounds a little strange (to me).. Also we have /users/ in the wiki and think we should try to keep to the wiki where possible. If we want to make a change to the wiki we can make a issue to ensure everyone is on board.

mk4111 commented:
Good morning @JMurphyWeb! :) I'm happy with currying functions too, I was just curious about why you decided to do it that way instead - like you said though, let's revisit this when we get around to refactoring.

RE: route names - I think that sounds like a really sound idea. I agree that it would seem a little odd to see '/people' relative to '/users' in the url, but I wondered if consistency throughout would be nice. As suggested, I think we should open a new issue on this to discuss which we should adopt, if not both! :)

iteles commented:
To confirm, is this ready for @nelsonic to review?

mk4111 commented:
@iteles yes please!

nelsonic commented on server/handlers.js:
```diff
@@ -7,86 +7,76 @@ const hash = new Hashids(process.env.HASHID_KEY);
 
 const handlers = {};
```
`const` > `var` (_i.e. zero advantage. you just typed more_)

mk4111 replied:
I can understand why that's the case and will change it! Thank you :)

nelsonic commented on server/handlers.js:
```diff
@@ -7,86 +7,76 @@ const hash = new Hashids(process.env.HASHID_KEY);
 
-handlers.activateUser = (request, reply) => {
+// not sure if this is used for anything else at the moment?
+handlers.serveFile = (request, reply) => {
```
if it isn't _tested_ remove it.

mk4111 replied:
Will do this too!

nelsonic commented on server/handlers.js:
```diff
@@ -7,86 +7,76 @@ const hash = new Hashids(process.env.HASHID_KEY);
+handlers.activatePrimaryUser = (request, reply) => {
   const hashedId = request.params.hashedId; // currently not hashed
   const userId = request.params.hashedId; // hash.decode(request.params.hashedId);
+  const redis = request.redis;
```
any _advantage_ to `const redis = request.redis;` ...?
you _type_ less by just using `request.redis` _twice_ than declaring another variable (_**const**ant_) and have one fewer line of code...

mk4111 replied:
No advantage at all - we thought it might read better but we realise we've declared it many a times, and are therefore happy to switch this out. Thank you for letting us know! :)

nelsonic commented on server/handlers.js:
```diff
@@ -136,3 +121,33 @@ handlers.login = (request, reply) => {
+  return JSON.stringify(orgUpdated);
+};
+
+const addPasswordToUser = (hashed, user) => {
```
any _advantage_ to this type of function declaration? (_as opposed to_) 
```
function addPasswordToUser (hashed, user) {
const addPasswordToUser = (hashed, user) => {
```
which one is _fewer characters to type_...?

mk4111 replied:
In all honesty, we hadn't considered this thoroughly, but now that we do, we see that it makes sense to do it this way, and to not assign functions to const/var. In general is it considered bad practise to store functions as const or var? We can see why that might be the case, but because we've seen examples of both we were wondering if you could let us know your thoughts on it. (Please!) Thank you @nelsonic ! :)

JMurphyWeb replied:
If we used `var` in place of `const` in your example @nelsonic wouldn't this solve the fewer characters problem but allow us to carry on using arrow functions and therefore be more consistent with the rest of the code (allowing us to take advantage of some of the benefits of arrows)?

nelsonic replied:
there are no _empirical_ "_benefits_" to using arrow functions other than they _look cool_... 
Function literals are simpler see: https://github.com/iteles/Javascript-the-Good-Parts-notes#function-literal and give you all the same advantages with _complete clarity of intent_ for the reader.

The situation where an *arrow function* is good for _compactness_ when you are in-lining an array method function. otherwise a _named_ function literal is _always_ better for debugging. 👍

nelsonic commented on server/joi.js:
```diff
@@ -9,7 +9,7 @@ validate.adminAddUser = {
     organisation_id: Joi.number().min(0).required(),
     user_type: ['primary', 'admin'],
-    submit: 'Submit'
+    submit: Joi.any().optional()
```
what _is_ the `submit` parameter...?

mk4111 replied:
It's the part of the payload, from a form.
Without this validation, we get the following error:
```
{
  statusCode: 400,
  error: "Bad Request",
  message: ""submit" is not allowed", 
  validation: {
     source: "payload",
     keys: [
       "submit"
     ]
  }
}
```

nelsonic replied:
Sorry, I don't "_get_" it. 😕 
We re passing in a _parameter_ to signify that a form was _submitted_...?

mk4111 replied:
Hi @nelsonic! Thank you for your reply :)
I didn't think we were passing in 'submit' as a parameter, but had to look into it a wee bit more.
Found that our HTML 'submit' input button had a name attribute, of value 'submit' --> this seems to have been the reason why the submit parameter was appearing in our form. This has now been removed and it seems to work fine now, ofcourse without the submit parameter validation. Sorry about this!

nelsonic commented on server/routes.js:
```diff
@@ -1,4 +1,4 @@
-const val = require('./joi.js');
+const validate = require('./joi.js');
```
avoid calling your validation (_model_) the name of the _library_ you are using.
instead call it the _name_ of the data you are validating e.g: `person.js`

mk4111 replied:
Really interesting point - we will divvy up the validation into types of data (person and org), and do exactly that. Thank you :)

nelsonic commented on test/index.js:
```diff
@@ -1,2 +1,4 @@
-require('./db.test.js');
-require('./server.test.js');
+require('./server.js');
```
remove this file. _completely_.
use a `glob` to run all the files in the `/test` directory. e.g: [`./node_modules/tape/bin/tape ./test/*.test.js`](https://github.com/dwyl/hapi-error/blob/89eb26a9588e3115ac962d8f69b88588fc46692a/package.json#L8) where your test files are named `test/your-descriptive-name.test.js`

mk4111 replied:
We remember our conversation around this, and both @JMurphyWeb and I agree that this is better practise - we'll do implement this right away! Thank you :)

mk4111 replied:
Hi @nelsonic - I'm finding it hard to find a glob that will read tests in the test directory, as some of our tests are split into subfolders, and others are at the root of the test folder... I've tried ```test/**/*.test.js``` but this only runs tests in the subfolders, and ignores those in the root of the test folder...  Could you please help me out?

nelsonic commented on test/mock-data.js:
```diff
@@ -0,0 +1,75 @@
+const mockData = {};
```
you're defining a _**const**ant_ and then _mutating_ it... 😕

mk4111 replied:
We see your argument and agree - we will change this.

nelsonic commented on test/server.js:
```diff
@@ -3,32 +3,35 @@ const tape = require('tape');
+  });
+};
+
+testingEndpoints();
```
what is the _advantage_ to wrapping your test in a `const` function and then calling it ...? 😕

mk4111 replied:
None at all, but I guess we did this because we wanted to keep the format to ES6. We (think we) now know better... This function will be removed, but we will keep this in mind for future situations.

nelsonic commented:
Hi @JMurphyWeb / @mk4111 left a few comments. please let me know if anything is unclear by responding in-line. thanks.

mk4111 commented:
Accidentally referenced #64 (the PR) instead of #83 (this issue for sorting out codecov).
This PR should now include codecov

iteles commented:
@mk4111 Does that mean it's ready for review? If so, please assign to @nelsonic .
If not, it might make it clearer to add an `in-progress` label or `not ready to review` in big letters in the description of the PR.

mk4111 commented:
Hi @iteles this is what we have so far. We've made the changes that @nelsonic suggested, and have merged in the PR with codecov ( #78 now closed).
We currently have some issues with codecov, whereby the coverage is _not_ running all the tests... @nelsonic would you know why that might be the case? The test in question is [primary_user_activates_account.test.js](https://github.com/emfoundation/ce100-app/blob/refactor-add-tests/test/primary-user-activates-account.test.js), where we do go to the endpoint `people/activate/{userId}` --> this doesn't seem to be reflected in the coverage... I might be missing something here but I can't see what that might be...

JMurphyWeb commented:
@iteles Marie said it was ready for review but she was having a problem with travis I think, which i believe she has gittered nelson about.

JMurphyWeb commented:
@mk4111 nice! what was the solution?

JMurphyWeb commented:
@mk4111 onfinish seems to run at the end of all tests in all files. Can we leave it as a teardown to be run at the end of each file?

mk4111 commented:
@JMurphyWeb the solution was to create a runner file, but only use it as an entry point for the coverage script, so I'm not sure if that's a fix that @nelsonic would be happy with...

mk4111 commented:
@JMurphyWeb I've made the changes that you asked for, so in teardown, we flushdb, and in .onFinish (which should only technically be run once) we have client.end(true) and server.stop(). onFinish is declared in both test files - we'll need to figure out a more appropriate place to put it later and to not repeat it either.

nelsonic commented on README.md:
```diff
@@ -1,2 +1,6 @@
 # ce100-app
+
+[![codecov](https://codecov.io/gh/emfoundation/ce100-app/branch/master/graph/badge.svg)](https://codecov.io/gh/emfoundation/ce100-app)
```
worth adding the travis badge and david-dm badges too ... https://github.com/dwyl/repo-badges 😉

JMurphyWeb replied:
#95

nelsonic commented on package.json:
```diff
@@ -4,25 +4,39 @@
     "hapi": "^14.2.0",
-    "joi": "^9.0.4"
+    "hapi-redis-connection": "^1.0.1",
+    "hashids": "^1.1.0",
```
using hashids so you can _decode_ ?

JMurphyWeb replied:
not being used anymore #95

nelsonic commented on package.json:
```diff
@@ -4,25 +4,39 @@
   },
   "devDependencies": {
+    "istanbul": "^0.4.4",
+    "codecov": "^1.0.1",
```
don't need the `codecov` module as it's installed by the `before_install:` in `.travis.yml` 👍

JMurphyWeb replied:
great #95

nelsonic commented on server/handlers.js:
```diff
@@ -0,0 +1,112 @@
+
+const bcrypt = require('bcrypt');
+
+var handlers = {};
```
please create a `Technical Task` to split out each handler into it's own file and require in-line in a `routes.js` file. 
this will reduce the "_churn_" and merge conflicts a single `handlers.js` file will _inevitably_ have.

JMurphyWeb replied:
https://github.com/emfoundation/ce100-app/issues/94

👍

nelsonic commented on server/person.js:
```diff
@@ -0,0 +1,33 @@
+const Joi = require('joi');
```
again having all "models" in a _single_ file will lead to merge conflicts... 
please add a `Technical Task` to split out into 3 separate files in a `/models` sub-folder ... thanks!

JMurphyWeb replied:
https://github.com/emfoundation/ce100-app/issues/94

nelsonic commented on server/routes.js:
```diff
@@ -0,0 +1,43 @@
+  }
+];
+
+module.exports = routes;
```
Good. 
`const` adds no value.
Otherwise good. 👍

nelsonic commented on server/server.js:
```diff
@@ -0,0 +1,31 @@
+  });
+});
+
+server.route(routes);
```
it's "OK" to do in-line requires e.g: 
```js
server.route('./routes.js');
```
reduces the number of variables and lines of code. 😉

JMurphyWeb replied:
#95

nelsonic commented on server/start.js:
```diff
@@ -0,0 +1,7 @@
+const server = require('./server.js');
```
server/start.js adds _no value_ it's just one more file ... please add the `server.start` to `server.js` in the next PR. thanks.

JMurphyWeb replied:
#95

nelsonic commented on templates/layout/default.html:
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
this is an _exact_ duplicate of `templates/default.html` do we _need_ _**both**_?

JMurphyWeb replied:
We just spotted this morning. Sorted in Harrys branch but not refactor tests yet. #95

nelsonic commented on test/admin-adds-primary-user.test.js:
```diff
@@ -0,0 +1,56 @@
+const tape = require('tape');
```
again, `const` adds no value...

JMurphyWeb replied:
ok #95

nelsonic commented:
@JMurphyWeb / @mk4111 code looks good. left a couple of comments. 
Please create technical tasks for splitting out handlers into their own files.

And avoid using `const` to declare variables unless there is an _unequivocal_ reason for it.
Its a poorly/inconsistently implemented feature of ES6 that adds no value and creates a _false_ sense of security.
Its _obvious_ to everyone on the team that `var test = require('tape');` _not_ going to be _re-assigned_ in the test file. 

I can go into a _lot_ more detail on `const` _if requested_ but it's beyond the "scope" of this PR... 😉

nelsonic commented:
![pretty close](https://cloud.githubusercontent.com/assets/194400/18280154/0793fb08-7450-11e6-861b-71928eab46bd.png)
